feat: 61 nnnnat textfield

nnnnat opened the pull request:
Resolves #61 
Impact: **major**  
Type: **feature**

## Issue
Create TextInput component

## Solution
We've decided to use [reacto-form](http://forms.dairystatedesigns.com/reacto-form/) to manage our form's state. Reacto form breaks apart what we were calling the `TextField` into 3 separate components, a `Field` component that wraps an input/errors and provides the label rendering, an `Input` component that renders the actual input field and a `ErrorsBlock` component that will display a field's error message.

Since we're following the reacto-form way of creating a form this PR is only for the `TextInput` which will be the RC equivalate to the reacto `Input` component but only for text values.

## Testing
The input works as a typical user would expect, and matches the designs in #18 

## Notes
The label, helper text, and error message will be part of upcoming PR for the Field and ErrorsBlock components. #72 and #73

aldeed commented:
@rymorgan regarding disabled, we chose to only support read-only rather than disabled. Is there any good reason to use disabled? In general the browser differences between the two are that read only can be tabbed to and selected while disabled cannot. I don't see a reason to use disabled but I could be missing something.

And yes, numeric / cc / password will be separate components built from this one later

rymorgan commented:
Totally agree with you. It should be tabbable for screen readers and such, so totally good with the read only decision. My comments were around it looking "disabled" so that the user knows they need to complete some other step for the field to become enabled.

aldeed commented:
OK I see now what you mean

nnnnat commented on src/styleguide/styles.css:
```diff
@@ -5,6 +5,10 @@
  */
 
+* {
+  box-sizing: border-box;
```
I've removed this from the global styles and added to the component's styles

nnnnat commented on src/defaultComponentTheme.js:
```diff
@@ -1,13 +1,125 @@
+// font weights
+const fontWeightLight = "200";
+const fontWeightRegular = "400";
+const fontWeightBold = "700";
```
I believe these are the correct weights from https://zpl.io/agGQ9o9. I went ahead and update the google fonts link in the index.html

rymorgan replied:
Yep, those are correct, but the color is not right, which is my bad b/c I did the colors after the text field and didn't go back and fix. It should be @cool-grey-500 for the text input default.

nnnnat commented on src/defaultComponentTheme.js:
```diff
@@ -1,13 +1,125 @@
+ */
+
+// font stack
+const fontStack = "Source Sans Pro, sans-serif";
```
Yeah kinda, I thought I might need it when I was creating those variables but ended up not using it. I'll delete for now

nnnnat commented:
@rymorgan  I updated a few styles so the focus styles will be visible in Chrome/Safari, I think the browser's focus what overlaying our focus styles so I turned them off.

I also had an older version of fontawesome so I've updated that and the icons I'm using for the valid state.

Can you create a GIF of the clear button not going away on blur so I can try to recreate the issue?

rymorgan commented:
@nnnnat Not seeing that issue anymore so seems fine now. The rest looks good.

aldeed commented on src/components/TextInput/v1/TextInput.md:
```diff
@@ -0,0 +1,294 @@
+
+Valid
+
+When information is required and/or needs to be formatted in a specific way the text input should be validated. When it’s successful the field shows a success state. Passing the `hasBeenValidated` prop to the input will enabled the valid state.
```
Let's be careful to say "input" rather than "field" everywhere since we'll have another one called Field.

Also, clarify the last line here:

The "valid" state appearance is used when the `errors` prop is empty and the `hasBeenValidated` prop is true. Typically an ancestor form component will do the work of setting these properly for you.

aldeed commented on src/components/TextInput/v1/TextInput.md:
````diff
@@ -0,0 +1,294 @@
+```
+Invalid
+
+When information is required and/or needs to be formatted in a specific way the text input should be validated. When a input can’t be validated the input is highlighted with error styling. Passing an `errors` array to the input to enabled the invalid state.
````
"Passing an `errors` array"

change to

"Pass a non-empty `errors` array"

And add this sentence after:

Typically an ancestor form component will do the work of setting `errors` properly for you.

aldeed commented on src/components/TextInput/v1/TextInput.md:
````diff
@@ -0,0 +1,294 @@
+```
+Read Only / Disabled
+
+A disabled input is used when an action needs to be taken in before a button can be enabled. A user us unable to input information into a disabled input. Passing the `isReadOnly` prop to enabled read only / disabled state.
````
"taken in" should be just "taken"

"a button" should be "the input"

"A user us" -> "A user is"

"Passing" -> "Pass"

aldeed commented on src/components/TextInput/v1/TextInput.md:
````diff
@@ -0,0 +1,294 @@
+</div>
+```
+###### Input Icons
+Text inputs have the ability to add a custom icon to the input to give extra content to the input. Text input icons has two available props to customize the input.
````
Change to:

You can add a custom icon to the right side of any text input and optionally provide a click handler for it. Use these two props:

aldeed commented on src/components/TextInput/v1/TextInput.md:
```diff
@@ -0,0 +1,294 @@
+All of the icon props will accept a few different nodes to create the input icon.
+ - React Component
+ - `<i/>` tag used for icon fonts
+ - `<svg>` tag for inline svg
```
Capitalize last SVG

aldeed commented on src/components/TextInput/v1/TextInput.md:
````diff
@@ -0,0 +1,294 @@
+  </div>
+</div>
+```
+All of the icon props will accept a few different nodes to create the input icon.
````
"a few different nodes" -> "a few different types of nodes"

aldeed commented on src/components/TextInput/v1/TextInput.md:
````diff
@@ -0,0 +1,294 @@
+```
+
+#### Multi line text field
+```jsx
````
Should have a sentence above here explaining that you pass `shouldAllowLineBreaks` to enable multi-line
